### ISSUE TITLE
[Gcal Persistence] Restore default config values when configured values removed from config

### ISSUE THREAD

**Issue by mhilbush** [openhab/openhab1-addons#5460]
In `services/gcal-persistence.cfg`, if I set the `offset` and/or `executescript` parameters to a value other than the default, those changes are correctly picked up by the binding after the deactivate/activate.  If I then comment them out, after the binding deactivate/activate, the old config values are not replaced with their defaults.

It appears to because the fields `executeScript` and `offset` in `GCalPersistenceService` are declared static [here](https://github.com/openhab/openhab1-addons/blob/master/bundles/persistence/org.openhab.persistence.gcal/src/main/java/org/openhab/persistence/gcal/internal/GCalPersistenceService.java#L71) and [here](https://github.com/openhab/openhab1-addons/blob/master/bundles/persistence/org.openhab.persistence.gcal/src/main/java/org/openhab/persistence/gcal/internal/GCalPersistenceService.java#L79)

Sample `gcal-persistence.cfg`.  Note the addition of the PID marker, which is needed to account for parameter deletions as described [here](https://github.com/htreu/smarthome/commit/dda15e92782d0772dbad155949d9d024e4061ae0#diff-86f0c90fe5b196c7e3a99bb29591270eR192)
```
pid:org.openhab.gcal-persistence
calendar_name=openHAB
#offset=3
#executescript=
```

**Comment by 9037568** [comment]
While it is strange they're declared static, that's not why the changes don't take.

Instead, it would appear to be because:
a) There's no updated() method implemented to allow changing settings after the service is started.
b) The [initialization code](https://github.com/openhab/openhab1-addons/blob/master/bundles/persistence/org.openhab.persistence.gcal/src/main/java/org/openhab/persistence/gcal/internal/GCalPersistenceService.java#L97) doesn't allow setting null values.

**Comment by mhilbush** [comment]
So those fields need to be static because they're accessed in the static class `SynchronizationJob`.

Why are you marking this `about-to-be-closed`?  Isn't this a problem that should be fixed?

Can it be fixed by initializing them to their defaults at the beginning of `activate` rather than doing it when they're declared?

**Comment by 9037568** [comment]
I'd argue it's working exactly as it was written to work. But I agree that the way it was written appears to be wrong.

Yes, we can fix it by updating the `activate` method, assuming that causes no bad behaviour. Is that something you'd like to create a PR for?

**Comment by mhilbush** [comment]
Between this behavior and the pid marker thing, I probably spent 2-3 hours on the forum trying to help someone sort out an issue they were having [here](https://community.openhab.org/t/presence-simulation-with-google-calendar-scheduler/19281/52).

I find the pid marker thing confounding.  It's so counter-intuitive to edit a config file and not have the change take effect -- ever -- even after a restart.  So, for example, if I comment out a line in `services/gcal-persistence.cfg`, without the pid marker at the top of the cfg file, the default value for that parameter will NEVER take effect (unless I know to delete `userdata/config/org/openhab/gcal-persistence.config`).

I can create a PR. I can even do it for gcal-persistence and gcal, since it's typical for both addons to be used together. I tested it and it works fine.  However, without the pid marker in the cfg file, I don't think it will do any good, since the system always will use the contents of `userdata/config/org/openhab/gcal-persistence.config`.

WDYT?

**Comment by 9037568** [comment]
I tend to agree with you. I dislike the entire shadow cache system for the reasons you highlight and others. I honestly think it should be done away with.

However, since that's not likely, this change would be a slight improvement. I'd be in favor of adopting this change if you want to submit it.

**Comment by mhilbush** [comment]
Now that the pid markers are (about to be) [documented](https://github.com/eclipse/smarthome/pull/4984) as a way to force the cfg file to be the exclusive service configuration, has there been any thought given to including pid markers in the OH1 addon cfg files [here](https://github.com/openhab/openhab1-addons/tree/b2f779c2a687b42912f68496686736b0c63c4414/features/openhab-addons-external/src/main/resources/conf)?

**Comment by 9037568** [comment]
First I'd heard of that. I would think, though, that if they were added to the cfg files, we'd want them to be commented out so that they aren't active by default.

**Comment by mhilbush** [comment]
@9037568 I did some additional testing on this.  The change I proposed above was supposed to fix the scenario where you comment out a parameter in the cfg file, the result of which is that the binding never reverts to the default value for that parameter. Unfortunately, without the pid marker, the change I suggested won't have any effect, as the parameter always will be pulled from the shadow config file in `userdata/config/org/openhab`.

Therefore, I'm going to close this issue. There's still a problem when commenting out config parameters, but I don't know how to fix it.
